csrc/scheduler/matmul_heuristic_plugin.cpp

```diff
@@ -0,0 +1,190 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <ir/interface_nodes.h>
+#include <mma_type.h>
+#include <scheduler/matmul_heuristic_plugin.h>
+#include <scheduler/mma_utils.h>
+#include <utils.h>
+
+#include <dlfcn.h>
+#include <cstdint>
+#include <mutex>
+
+namespace nvfuser {
+
+namespace matmul_heuristic_plugin {
+
+namespace {
+
+//! Defines ConfigFactoryFuncPtr as type of the "makeConfig" symbol
+typedef KernelConfig* (*ConfigFactoryFuncPtr)();
+
+std::mutex plugin_mutex;
+class PluginInterface : NonCopyable {
+ public:
+  PluginInterface() : filepath_(getNvFuserEnv("MATMUL_HEURISTIC_PLUGIN")) {
+    if (filepath_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      library_handle_ = dlopen(filepath_, RTLD_LAZY);
+      NVF_CHECK(
+          library_handle_ != nullptr,
+          "Error occurred when loading matmul heuristic plugin ",
+          filepath_,
+          ". Error msg: ",
+          dlerror());
+    }
+  }
+
+  bool available() const {
+    return library_handle_ != nullptr;
+  }
+
+  ~PluginInterface() {
+    if (library_handle_ != nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      dlclose(library_handle_);
+      library_handle_ = nullptr;
+      factory_func_ptr_ = nullptr;
+    }
+  }
+
+  KernelConfig* makeConfig() {
+    NVF_ERROR(available());
+
+    if (factory_func_ptr_ == nullptr) {
+      std::lock_guard<std::mutex> lock(plugin_mutex);
+      factory_func_ptr_ =
+          (ConfigFactoryFuncPtr)dlsym(library_handle_, "makeConfig");
+      NVF_CHECK(
+          factory_func_ptr_ != nullptr,
+          "Failed to load symbol \"makeConfig\" from plugin file ",
+          filepath_,
+          ". Error message: ",
+          dlerror());
+    }
+
+    return (*factory_func_ptr_)();
+  }
+
+ private:
+  char* filepath_ = nullptr;
+  void* library_handle_ = nullptr;
+  ConfigFactoryFuncPtr factory_func_ptr_ = nullptr;
```

Comment:
Should we pull this out of `PluginInterface` and add a setter `setMatmulConfigFactory` so that we can provide the factory function from a different source? For example, in a unit test, we can
```C++
TEST_F(NVFuserTest, TestMatmulHeuristicsPlugin){ 
  struct MyTestPlugin : public matmul_heuristic_plugin::KernelConfig {
    ....
    void configure() override {
      EXPECT_TRUE("the problem desc we get matches with what we set");
      set kernel config
    }
  };

  auto f = [&executed]() { return new MyTestPlugin() };
  setMatmulConfigFactory(f);
  ...
  EXPECT_TRUE("the kernel config we get matches with what we set");
}
```
we can even set the factory function as a python function.

Reply:
Good idea. Actually I wonder an `AbstractPluginInterface` class should be declared public with a virtual `makeConfig`. We could then accept a `AbstractPluginInterface* plugin_interface = nullptr` argument in `updateMatmulParams()`. It would just be a little more work then to implement a test I think:
```c++
TEST_F(NVFuserTest, TestMatmulHeuristicsPlugin){ 
  struct MyTestPlugin : public matmul_heuristic_plugin::KernelConfig {
    ....
    void configure() override {
      EXPECT_TRUE("the problem desc we get matches with what we set");
      set kernel config
    }
  };
  class MyPluginInterface : public AbstractPluginInterface {
    KernelConfig* makeConfig() override {
      return static_cast<KernelConfig*>(new MyTestPlugin);
    }
  } test_plugin;

  updateMatmulParams(..., /*plugin_interface=*/test_plugin);
  ...
  EXPECT_TRUE("the kernel config we get matches with what we set");
}
```

Reply:
I see what you mean about python functions though. With my approach in the last comment, it would be awkward to plumb in a python heuristic, compared to being able to just set it globally as in your example.

Reply:
Yeah, the point is to have some context specifying that function, and the ability to update this context.